onnx/defs/parser.h

```diff
@@ -0,0 +1,377 @@
+    return Status::OK();
+  }
+
+  Status Parse(uint64_t& val) {
+    Literal literal;
+    CHECK_PARSER_STATUS(Parse(literal));
+    if (literal.type != LiteralType::INT_LITERAL)
+      return ParseError("Integer value expected, but not found.");
+    std::string s = literal.value;
+    val = std::stoull(s);
+    return Status::OK();
+  }
+
+  Status Parse(float& val) {
+    Literal literal;
+    CHECK_PARSER_STATUS(Parse(literal));
+    switch (literal.type) {
+      case LiteralType::INT_LITERAL:
+      case LiteralType::FLOAT_LITERAL:
+        val = std::stof(literal.value);
+        break;
+      default:
+        return ParseError("Unexpected literal type.");
+    }
+    return Status::OK();
+  }
+
+  Status Parse(double& val) {
+    Literal literal;
+    CHECK_PARSER_STATUS(Parse(literal));
+    switch (literal.type) {
+      case LiteralType::INT_LITERAL:
+      case LiteralType::FLOAT_LITERAL:
+        val = std::stod(literal.value);
+        break;
+      default:
+        return ParseError("Unexpected literal type.");
+    }
+    return Status::OK();
+  }
+
+  // Parse a string-literal enclosed within doube-quotes.
+  Status Parse(std::string& val) {
+    Literal literal;
+    CHECK_PARSER_STATUS(Parse(literal));
+    if (literal.type != LiteralType::STRING_LITERAL)
+      return ParseError("String value expected, but not found.");
+    val = literal.value;
+    return Status::OK();
+  }
+
+  // Parse an identifier, including keywords. If none found, this will
+  // return an empty-string identifier.
+  Status ParseOptionalIdentifier(std::string& id) {
+    SkipWhiteSpace();
+    auto from = next_;
+    if ((next_ < end_) && (isalpha(*next_) || (*next_ == '_'))) {
+      ++next_;
+      while ((next_ < end_) && (isalnum(*next_) || (*next_ == '_')))
+        ++next_;
+    }
+    id = std::string(from, next_ - from);
+    return Status::OK();
+  }
+
+  Status ParseIdentifier(std::string& id) {
+    ParseOptionalIdentifier(id);
+    if (id.empty())
+      return ParseError("Identifier expected but not found.");
+    return Status::OK();
+  }
+
+  Status PeekIdentifier(std::string& id) {
+    SavePos();
+    ParseOptionalIdentifier(id);
+    RestorePos();
+    return Status::OK();
+  }
+
+  Status Parse(KeyWordMap::KeyWord& keyword) {
+    std::string id;
+    CHECK_PARSER_STATUS(ParseIdentifier(id));
+    keyword = KeyWordMap::Lookup(id);
+    return Status::OK();
+  }
+
+ protected:
+  const char* start_;
+  const char* next_;
+  const char* end_;
+  const char* saved_pos_;
+};
+
+class OnnxParser : public ParserBase {
+ public:
+  OnnxParser(const char* cstr) : ParserBase(cstr) {}
+
+  Status Parse(TensorShapeProto& shape);
+
+  Status Parse(TypeProto& typeProto);
+
+  Status Parse(TensorProto& tensorProto);
+
+  Status Parse(AttributeProto& attr);
+
+  Status Parse(AttrList& attrlist);
+
+  Status Parse(NodeProto& node);
+
+  Status Parse(NodeList& nodelist);
+
+  Status Parse(GraphProto& graph);
+
+  Status Parse(ModelProto& model);
```

Comment:
I'm still not sure why we don't hide all of these `Parse(...Proto&)` methods as `private`

Reply:
(a) My experience has been that developers construct the different kinds of objects (such as a TensorShapeProto, or TypeProto) frequently and a convenient way to construct them is useful.
(b) Some of the function-definitions require constructing a graph in which some parts are dependent on context-information (like an input type T); the flexibility exposed by the above interface will make this easier.
(c) A more minor point is that it is convenient to have unit-test-cases that test parsing of the different components. (We could possibly address these using friend-classes, but given the other reasons, the above seems simpler.)
In summary, I think the advantages of the above outweigh the disadvantages (if any). (Of course, we can revisit this later, but unless we expose this functionality, there is no way for users to test it or try it out and give feedback.)